### PR TITLE
Various tests + Poison Heal fix + Transistor update

### PR DESCRIPTION
**Pull request by Bassoonian**
Replaces #3181 

## Description
Adds battle tests for Defeatist, Gale Wings, Poison Heal, Sap Sipper, Earth Eater, Steelworker, Transistor, Dragon's Maw and Rocky Payload. Fixes Poison Heal ability popup and adds a config for Transistor's Gen9 nerf.

## **Discord contact info**
bassoonian

### CODE REVIEW

File: test/battle/ability/defeatist.c [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_ECHOED_VOICE); MOVE(opponent, MOVE_CELEBRATE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_ECHOED_VOICE, player);
+        HP_BAR(opponent, captureDamage: &results[i].damage);
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_CELEBRATE, opponent);
+    } FINALLY {
+        EXPECT_MUL_EQ(results[0].damage, Q_4_12(0.5), results[1].damage);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/battle/ability/defeatist.c:46]: New line at the end of the file

File: test/battle/ability/gale_wings.c [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        OPPONENT(SPECIES_WOBBUFFET) { Speed(100);};
+    } WHEN {
+        TURN { MOVE(player, MOVE_AERIAL_ACE); }
+    } SCENE {
+        if (hp == 100) {
+            MESSAGE("Talonflame used Aerial Ace!");
+        }
+        else {
+            MESSAGE("Foe Wobbuffet used Celebrate!");
+        }
```
Comment (AsparagusEduardo) [inline test/battle/ability/gale_wings.c:22]: This doesn't properly show that one message happens after the other.
A single `MESSAGE` just checks if the message happens at all.
```suggestion
        if (hp == 100) {
            MESSAGE("Talonflame used Aerial Ace!");
            MESSAGE("Foe Wobbuffet used Celebrate!");
        }
        else {
            MESSAGE("Talonflame used Aerial Ace!");
            MESSAGE("Foe Wobbuffet used Celebrate!");
        }
```

File: test/battle/ability/gale_wings.c [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        OPPONENT(SPECIES_WOBBUFFET) { Speed(100);};
+    } WHEN {
+        TURN { MOVE(player, move); }
+    } SCENE {
+        if (move == MOVE_AERIAL_ACE) {
+            MESSAGE("Talonflame used Aerial Ace!");
+        }
+        else {
+            MESSAGE("Foe Wobbuffet used Celebrate!");
+        }
```
Comment (AsparagusEduardo) [inline test/battle/ability/gale_wings.c:45]: Same as above.
```suggestion
        if (move == MOVE_AERIAL_ACE) {
            MESSAGE("Talonflame used Aerial Ace!");
            MESSAGE("Foe Wobbuffet used Celebrate!");
        }
        else {
            MESSAGE("Foe Wobbuffet used Celebrate!");
            MESSAGE("Talonflame used Aerial Ace!");
        }
```

File: test/battle/ability/gale_wings.c [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+}
```
Comment (AsparagusEduardo) [inline test/battle/ability/gale_wings.c:47]:
```suggestion
}

TO_DO_BATTLE_TEST("Gale Wings doesn't increase priority of Flying-type Hidden Power, Natural Gift, Judgment or Tera Blast");
```

File: test/battle/ability/poison_heal.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    GIVEN {
+        PLAYER(SPECIES_SHROOMISH) { Ability(ABILITY_POISON_HEAL); Status1(STATUS1_POISON);  HP(1), MaxHP(400); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_HEAL_BLOCK); }
+    } SCENE {
+        NONE_OF {
+            ABILITY_POPUP(player, ABILITY_POISON_HEAL);
+            MESSAGE("The poisoning healed Shroomish a little bit!");
+            HP_BAR(player, damage: -50);
```
Comment (AsparagusEduardo) [inline test/battle/ability/poison_heal.c:43]:
```suggestion
            HP_BAR(player, damage: -50);
            HP_BAR(player, damage: 50);
```

File: test/battle/ability/poison_heal.c [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_CELEBRATE); }
+    } SCENE {
+        ABILITY_POPUP(player, ABILITY_POISON_HEAL);
+        MESSAGE("The poisoning healed Shroomish a little bit!");
+        HP_BAR(player, damage: -50);
+    }
+}
+
+SINGLE_BATTLE_TEST("Poison Heal heals from Toxic Poison damage")
```
Comment (AsparagusEduardo) [inline test/battle/ability/poison_heal.c:18]: This test should account for the fact that Toxic Poison heals at the same rate as regular Poison.

File: test/battle/ability/poison_heal.c [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    } WHEN {
+        TURN { MOVE(player, MOVE_CELEBRATE); }
+    } SCENE {
+        NONE_OF {
+            ABILITY_POPUP(player, ABILITY_POISON_HEAL);
+            MESSAGE("The poisoning healed Shroomish a little bit!");
+            HP_BAR(player, damage: -50);
+        }
+    }
+}
```
Comment (AsparagusEduardo) [inline test/battle/ability/poison_heal.c:62]: New line at the end of file.

File: test/battle/ability/sap_sipper.c [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        PLAYER(SPECIES_MARILL) { Ability(ABILITY_SAP_SIPPER); }
+        OPPONENT(SPECIES_WOBBUFFET) { Speed(1); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_BELLY_DRUM); MOVE(opponent, MOVE_VINE_WHIP); }
+    } SCENE {
+        ABILITY_POPUP(player, ABILITY_SAP_SIPPER);
+        NONE_OF { ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player); }
+        NONE_OF { MESSAGE("Marill's Attack rose!"); }
+    }
+}
```
Comment (AsparagusEduardo) [inline test/battle/ability/sap_sipper.c:55]: New line at the end of file.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Comment by Bassoonian** [comment]
Ready for re-review. Also added tests for Earth Eater as well as Steelworker and its variants, and added a config for Transistor (which is nerfed in Gen9+).

**Comment by AsparagusEduardo** [comment]
> Ready for re-review. Also added tests for Earth Eater as well as Steelworker and its variants, and added a config for Transistor (which is nerfed in Gen9+).

Update the description and title with the new changes.

**Comment by Bassoonian** [comment]
Done!

File: src/battle_util.c [hunk 9]
```diff
@@ -8964,7 +8964,11 @@ static inline u32 CalcMoveBasePowerAfterModifiers(u32 move, u32 battlerAtk, u32
         break;
     case ABILITY_TRANSISTOR:
         if (moveType == TYPE_ELECTRIC)
+        #if B_TRANSISTOR >= GEN_9
+            modifier = uq4_12_multiply(modifier, UQ_4_12(5325 / 4096));
```
Comment (mrgriffin) [inline src/battle_util.c:8968]: Damn, this isn't quite `UQ_4_12(1.3)`—`1.3 * 4096` is `5324.8` which truncates down to `5324`. I wonder if we could change it to round :thinking:

Anyway, I won't ask you to do that! Just something for me to look at for a future PR :)

**Review by mrgriffin — APPROVED** [review]

File: src/battle_util.c (hunk 9 above)
Comment (Bassoonian) [reply src/battle_util.c:8968]: I did try 1.3 while having the actual division in the code, but the test failed so having it here was the only way to make it pass :( it's unfortunate indeed, but we could always just take artistic license and simplify it to 1.3

File: test/battle/ability/poison_heal.c [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+{
+    s16 turnOneHit;
+    s16 turnTwoHit;
+
+    GIVEN {
+        PLAYER(SPECIES_SHROOMISH) { Ability(ABILITY_POISON_HEAL); Status1(STATUS1_TOXIC_POISON);  HP(1), MaxHP(400); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_CELEBRATE); }
+        TURN { MOVE(player, MOVE_CELEBRATE); }
```
Comment (AsparagusEduardo) [inline test/battle/ability/poison_heal.c:32]:
```suggestion
        TURN { }
        TURN { }
```

File: test/battle/ability/poison_heal.c [hunk 11]
```diff
@@ -0,0 +1,77 @@
+#include "global.h"
+#include "test/battle.h"
+
+SINGLE_BATTLE_TEST("Poison Heal heals from (Toxic) Poison damage")
+{
+    u8 status;
+    PARAMETRIZE { status=STATUS1_POISON; }
+    PARAMETRIZE { status=STATUS1_TOXIC_POISON; }
```
Comment (AsparagusEduardo) [inline test/battle/ability/poison_heal.c:8]:
```suggestion
    PARAMETRIZE { status = STATUS1_POISON; }
    PARAMETRIZE { status = STATUS1_TOXIC_POISON; }
```

File: test/battle/ability/gale_wings.c [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+SINGLE_BATTLE_TEST("Gale Wings only grants priority to Flying-type moves")
+{
+    u32 move;
+    PARAMETRIZE { move = MOVE_AERIAL_ACE; }
+    PARAMETRIZE { move = MOVE_FLARE_BLITZ; }
+    GIVEN {
+        ASSUME(B_GALE_WINGS >= GEN_7);
+        ASSUME(gBattleMoves[MOVE_AERIAL_ACE].type == TYPE_FLYING);
+        ASSUME(gBattleMoves[MOVE_FLARE_BLITZ].type == TYPE_FIRE);
```
Comment (AsparagusEduardo) [inline test/battle/ability/gale_wings.c:36]:
```suggestion
        ASSUME(gBattleMoves[MOVE_FLARE_BLITZ].type != TYPE_FLYING);
```

File: test/battle/ability/dragons_maw.c [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+    PARAMETRIZE { move = MOVE_TACKLE; ability = ABILITY_KLUTZ; }
+    PARAMETRIZE { move = MOVE_TACKLE; ability = ABILITY_DRAGONS_MAW; }
+    PARAMETRIZE { move = MOVE_DRAGON_CLAW; ability = ABILITY_KLUTZ; }
+    PARAMETRIZE { move = MOVE_DRAGON_CLAW; ability = ABILITY_DRAGONS_MAW; }
+    PARAMETRIZE { move = MOVE_DRAGON_BREATH; ability = ABILITY_KLUTZ; }
+    PARAMETRIZE { move = MOVE_DRAGON_BREATH; ability = ABILITY_DRAGONS_MAW; }
+
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_TACKLE].type == TYPE_NORMAL);
```
Comment (AsparagusEduardo) [inline test/battle/ability/dragons_maw.c:17]:
```suggestion
        ASSUME(gBattleMoves[MOVE_TACKLE].type != TYPE_DRAGON);
```

File: src/battle_util.c (hunk 9 above)
Comment (AsparagusEduardo) [reply src/battle_util.c:8968]: I want to see where this goes before accepting it.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Comment by Bassoonian** [comment]
Should be good to go other than the 5325/4096 multiplier, which I'll leave for @AsparagusEduardo to figure out :)

**Review by AsparagusEduardo — APPROVED** [review]

**Comment by AsparagusEduardo** [comment]
*crap, the transistor update should've gone to `upcoming`*

**Pull request merged**